src/coreComponents/finiteVolume/CellElementStencilTPFA.hpp

```diff
@@ -97,8 +164,108 @@ class CellElementStencilTPFA : public StencilBase< CellElementStencilTPFA_Traits
     return MAX_STENCIL_SIZE;
   }
 
+  /// Type of kernel wrapper for in-kernel update
+  using StencilWrapper = CellElementStencilTPFAWrapper;
+
+  /**
+   * @brief Create an update kernel wrapper.
+   * @return the wrapper
+   */
+  StencilWrapper createStencilWrapper()
+  {
+    return StencilWrapper( m_elementRegionIndices,
+                           m_elementSubRegionIndices,
+                           m_elementIndices,
+                           m_weights,
+                           m_faceNormal,
+                           m_cellToFaceVec,
+                           m_transMultiplier );
+  }
+
+private:
+  array2d< real64 > m_faceNormal;
+  array3d< real64 > m_cellToFaceVec;
+  array1d< real64 > m_transMultiplier;
+
 };
 
+template< typename PERMTYPE >
+void CellElementStencilTPFAWrapper::computeTransmissibility( localIndex iconn,
+                                                             PERMTYPE permeability,
+                                                             real64 (& transmissibility)[2] ) const
+{
+  real64 halfTrans[2];
+
+  // real64 const tolerance = 1e-30 * lengthTolerance; // TODO: choice of constant based on physics?
+
+  for( localIndex i =0; i<2; i++ )
+  {
+    localIndex const er  =  m_elementRegionIndices[iconn][i];
+    localIndex const esr =  m_elementSubRegionIndices[iconn][i];
+    localIndex const ei  =  m_elementIndices[iconn][i];
+
+    halfTrans[i] = m_weights[iconn][i];
+
+    // Proper computation
+    real64 faceNormal[3], faceConormal[3];
+
+    LvArray::tensorOps::copy< 3 >( faceNormal, m_faceNormal[iconn] );
+
+    if( LvArray::tensorOps::AiBi< 3 >( m_cellToFaceVec[iconn][i], faceNormal ) < 0.0 )
+    {
+      LvArray::tensorOps::scale< 3 >( faceNormal, -1 );
+    }
+
+    LvArray::tensorOps::hadamardProduct< 3 >( faceConormal, permeability[er][esr][ei][0], faceNormal );
+    halfTrans[i] *= LvArray::tensorOps::AiBi< 3 >( m_cellToFaceVec[iconn][i], faceConormal );
+
+    // correct negative weight issue arising from non-K-orthogonal grids
+    if( halfTrans[i] < 0.0 )
+    {
+      LvArray::tensorOps::hadamardProduct< 3 >( faceConormal,
+                                                permeability[er][esr][ei][0],
+                                                m_cellToFaceVec[iconn][i] );
+      halfTrans[i] = m_weights[iconn][i];
+      halfTrans[i] *= LvArray::tensorOps::AiBi< 3 >( m_cellToFaceVec[iconn][i], faceConormal );
+    }
+  }
+
+  // Do harmonic and arithmetic averaging
+
+  real64 const harmonicWeight   = halfTrans[0]*halfTrans[1] / (halfTrans[0]+halfTrans[1]);
+  real64 const arithmeticWeight = ( halfTrans[0]+halfTrans[1] )/ 2;
+
+  real64 const meanPermCoeff = 1.0; //TODO make it a member if it is really necessary
+
+  real64 const value = meanPermCoeff * harmonicWeight + (1 - meanPermCoeff) * arithmeticWeight;
+  for( localIndex ke = 0; ke < 2; ++ke )
+  {
+    transmissibility[ke] = m_transMultiplier[iconn] * value * (ke == 0 ? 1 : -1);
+  }
+}
```

Comment:
Would it be feasible/desirable/easier to compute the derivatives with respect to pressure in this function, instead of in `dTrans_dPressure`, since you will need the half-transmissibilities to compute them?

Also, do you plan to keep the averaging between `harmonicWeight` and `arithmeticWeight` for the `CellElementSubRegion`?

Reply:
1. Yes, it would be certainly feasible. I kind of like the idea to keep functions to do only one task and I also thought that eventually we will need a functions to compute `dTrans_dDisp` and so I thought it would be cleaner to have 2 separate functions. Putting everything in a single function would not be hard though.

2. I am no sure. I saw that the arithmetic av. was there for the fractures and so I kept it but I m not sure why someone would want to use it. If we think it s useless I don't mind removing it.